File: 19th/코딩테스트 심화반/week2/18111번_마인크래프트.cpp
#include <iostream>

using namespace std;

// 전역 변수에 선언하면 모든 값이 0으로 초기화 됨
int map[501][501];

int main() {
    int N, M, B;
    
    cin >> N >> M >> B;
    
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cin >> map[i][j];
        }
    }
    
    int min_time = 1234567890; // 문제에서 나오지 않을 큰 수로 초기화
    int max_height = -1; // 문제에서 나오지 않을 작은 수로 초기화
    
    // 현재 높이 k를 0부터 256까지 설정하며 해당 높이로 블록을 쌓거나 제거하는데 걸리는 시간을 구함
    // 가장 시간이 짧고, 땅의 높이가 높은 선택지를 선택
    for (int k = 0; k <= 256; k++) {
        // 좌표 (i, j)의 가장 위에 있는 블록을 제거하여 인벤토리에 넣는 작업
        int job_one_cnt = 0; 
        // 인벤토리에서 블록 하나를 꺼내어 좌표 (i, j)의 가장 위에 있는 블록 위에 놓는 작업
        int job_two_cnt = 0;
        
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                // 현재 높이 k보다 높을 경우
                // 블록을 현재 높이 k로 맞추기 위해 map[i][j] - k만큼 1번 작업 수행
                if (k < map[i][j]) job_one_cnt += map[i][j] - k; 

                // 현재 높이 k보다 낮을 겨우
                // 블록을 현재 높이 k로 맞추기 위해 k - map[i][j]만큼 2번 작업 수행
                else if (k > map[i][j]) job_two_cnt += k - map[i][j];
            }
        }
        
        // 초기 블록 개수 + 1번 작업을 통해 얻은 블록의 개수가 사용한 블록의 개수보다 같거나 많아야 가능
        if (B + job_one_cnt - job_two_cnt >= 0) { 
            
            // 1번 작업 * 2 + 2번 작업의 시간이 현재 최소 시간보다 작거나 같은 경우
            if (min_time >= job_one_cnt * 2 + job_two_cnt) {
                // 최소 시간 갱신
                min_time = job_one_cnt * 2 + job_two_cnt;
                // 최대 높이 갱신
                max_height = k;
            }
        }
    }
    
    cout << min_time << " " << max_height;
}

File: 20th/코딩테스트 기초반/week2/18111번_마인크래프트.py
n, m, b = map(int, input().split())

block = []
for _ in range(n):
    block.append(list(map(int, input().split())))

min_time = 1e9  # 최소 시간을 아주 큰 값으로 초기화
max_height = -1  # 가장 높은 높이를 저장할 변수

# 가능한 높이(0 ~ 256) 모두 시도
for height in range(257):
    job_one_cnt = 0  # 작업1 수
    job_two_cnt = 0  # 작업2 수

    # 모든 좌표 순회
    for i in range(n):
        for j in range(m):
            if block[i][j] > height:  # 현재 높이가 목표 높이보다 높으면 블록 제거 
                job_one_cnt += (block[i][j] - height)
            elif block[i][j] < height:  # 현재 높이가 목표 높이보다 낮으면 블록 설치 
                job_two_cnt += (height - block[i][j])
    
    # 인벤토리 블록으로 작업이 가능한 경우 (남는 블록 + 인벤토리 블록으로 부족한 만큼 채울 수 있는지 확인)
    if b - (job_two_cnt - job_one_cnt) >= 0:
        if min_time >= (job_one_cnt * 2 + job_two_cnt):
            min_time = job_one_cnt * 2 + job_two_cnt
            max_height = height

print(min_time, max_height)